In [54]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

In [55]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))
train_df

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0         7000.0    0.1894       256.38       MORTGAGE     75000.0   
1        40000.0    0.1614       975.71       MORTGAGE    102000.0   
2        11000.0    0.2055       294.81           RENT     45000.0   
3         4000.0    0.1612       140.87       MORTGAGE     38000.0   
4        14000.0    0.1797       505.93       MORTGAGE     43000.0   
...          ...       ...          ...            ...         ...   
12175    19975.0    0.2565       801.09           RENT     28000.0   
12176    15000.0    0.1774       540.34           RENT     50000.0   
12177     3600.0    0.1862       131.28           RENT     60000.0   
12178    15000.0    0.0881       475.68       MORTGAGE     62000.0   
12179    15000.0    0.1774       540.34           RENT     54080.0   

      verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0            Not Verified          n  28.62          0.0             2.0  ...   
1         Source Verified          n  11.72          2.0             0.0  ...   
2                Verified          n  37.25          1.0             3.0  ...   
3            Not Verified          n  42.89          1.0             0.0  ...   
4         Source Verified          n  22.16          1.0             0.0  ...   
...                   ...        ...    ...          ...             ...  ...   
12175        Not Verified          n  28.42          0.0             0.0  ...   
12176            Verified          n  23.43          4.0             0.0  ...   
12177        Not Verified          n  28.80          0.0             1.0  ...   
12178     Source Verified          n  11.44          0.0             0.0  ...   
12179        Not Verified          n  21.66          1.0             1.0  ...   

       percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0                  87.5                   0.0        0.0         352260.0   
1                   0.0                   0.0        0.0         294664.0   
2                   7.7                   0.0        0.0          92228.0   
3                 100.0                   0.0        0.0         284273.0   
4                  25.0                   0.0        0.0         120280.0   
...                 ...                   ...        ...              ...   
12175              16.7                   0.0        0.0          50055.0   
12176              11.1                   0.0        0.0          70324.0   
12177               0.0                   0.0        0.0          83765.0   
12178               0.0                   0.0        0.0         189930.0   
12179             100.0                   0.0        0.0          48968.0   

      total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0               62666.0         35000.0                     10000.0   
1              109911.0          9000.0                     71044.0   
2               36007.0         33000.0                     46328.0   
3               52236.0         13500.0                     52017.0   
4               88147.0         33300.0                     78680.0   
...                 ...             ...                         ...   
12175           28192.0         18700.0                     19055.0   
12176           57025.0         13300.0                     54824.0   
12177           55156.0         14800.0                     53065.0   
12178           23748.0          7000.0                     32930.0   
12179           29786.0          2900.0                     33968.0   

       hardship_flag  debt_settlement_flag     target  
0                  N                     N   low_risk  
1                  N                     N   low_risk  
2                  N                     N   low_risk  
3                  N                     N   low_risk  
4                  N                     N   low_risk  
...              ...                   ...        ...  
12175              N   

In [56]:
test_df

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0       40000.0    0.1033       856.40           RENT    128700.0   
1       24450.0    0.1430       572.72       MORTGAGE     44574.0   
2       13500.0    0.1430       316.23            OWN     60000.0   
3       10625.0    0.1774       268.31           RENT     60000.0   
4        6375.0    0.1862       232.46           RENT     60000.0   
...         ...       ...          ...            ...         ...   
4697    30000.0    0.1240       673.42           RENT    140480.0   
4698    24000.0    0.0756       747.22           RENT     50000.0   
4699    10000.0    0.2305       387.36           RENT     33000.0   
4700     8000.0    0.1862       205.86           RENT     38000.0   
4701    30000.0    0.2055      1123.34           RENT    180000.0   

     verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Source Verified          n  12.47          0.0             1.0  ...   
1           Not Verified          n  15.05          0.0             1.0  ...   
2           Not Verified          n  28.72          0.0             0.0  ...   
3               Verified          n  15.70          0.0             4.0  ...   
4        Source Verified          n  35.50          0.0             0.0  ...   
...                  ...        ...    ...          ...             ...  ...   
4697     Source Verified          n  15.74          0.0             0.0  ...   
4698        Not Verified          n  26.81          0.0             0.0  ...   
4699            Verified          n  38.51          0.0             2.0  ...   
4700     Source Verified          n  16.36          0.0             1.0  ...   
4701     Source Verified          n  12.06          0.0             0.0  ...   

      percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0                 57.1                   0.0        0.0          63915.0   
1                  0.0                   0.0        0.0         136425.0   
2                  0.0                   0.0        0.0          82124.0   
3                 20.0                   0.0        0.0          54855.0   
4                 75.0                   0.0        0.0          90445.0   
...                ...                   ...        ...              ...   
4697              28.6                   0.0        0.0         159688.0   
4698               0.0                   0.0        0.0          62375.0   
4699               0.0                   0.0        0.0          43250.0   
4700               0.0                   1.0        0.0          31357.0   
4701              33.3                   0.0        0.0         218686.0   

     total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0              49510.0         49400.0                     14515.0   
1              19439.0         15500.0                     18925.0   
2              65000.0          5400.0                     61724.0   
3              50335.0         23200.0                     26255.0   
4              56541.0         15300.0                     72345.0   
...                ...             ...                         ...   
4697          110873.0         48400.0                    107388.0   
4698           18928.0         13300.0                     30775.0   
4699           33022.0          8500.0                     29550.0   
4700           19595.0          1500.0                      9657.0   
4701          209389.0          5600.0                    209986.0   

      hardship_flag  debt_settlement_flag     target  
0                 Y                     N   low_risk  
1                 N                     N   low_risk  
2                 Y                     N   low_risk  
3                 N                     N   low_risk  
4                 N                     N   low_risk  
...             ...                   ...        ...  
4697              N                     N  high_risk  
4698              N   

In [57]:
# Convert categorical data to numeric and separate target feature for training data
y_train = train_df["target"]
X_train = train_df.drop(columns = ["target"])
X_train = pd.get_dummies(X_train)
X_train.head()

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0     7000.0    0.1894       256.38     75000.0  28.62          0.0   
1    40000.0    0.1614       975.71    102000.0  11.72          2.0   
2    11000.0    0.2055       294.81     45000.0  37.25          1.0   
3     4000.0    0.1612       140.87     38000.0  42.89          1.0   
4    14000.0    0.1797       505.93     43000.0  22.16          1.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             2.0      20.0      0.0    40414.0  ...   
1             0.0      10.0      0.0    43531.0  ...   
2             3.0      23.0      0.0     8242.0  ...   
3             0.0       7.0      0.0    12767.0  ...   
4             0.0      22.0      0.0    11182.0  ...   

   verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                             0             1                      1   
1                             0             1                      0   
2                             1             1                      0   
3                             0             1                      0   
4                             0             1                      0   

   initial_list_status_w  application_type_Individual  \
0                      0                            1   
1                      1                            1   
2                      1                            1   
3                      1                            0   
4                      1                            1   

   application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                           0                1                0   
1                           0                1                0   
2                           0                1                0   
3                           1                1                0   
4                           0                1                0   

   debt_settlement_flag_N  debt_settlement_flag_Y  
0                       1                       0  
1                       1                       0  
2                       1                       0  
3                       1                       0  
4                       1                       0  

[5 rows x 92 columns]

In [58]:
# Convert categorical data to numeric and separate target feature for testing data
y_test = test_df["target"]
X_test = test_df.drop(columns = ["target"])
X_test = pd.get_dummies(X_test)
X_test.head()

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    40000.0    0.1033       856.40    128700.0  12.47          0.0   
1    24450.0    0.1430       572.72     44574.0  15.05          0.0   
2    13500.0    0.1430       316.23     60000.0  28.72          0.0   
3    10625.0    0.1774       268.31     60000.0  15.70          0.0   
4     6375.0    0.1862       232.46     60000.0  35.50          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             1.0       8.0      0.0    38113.0  ...   
1             1.0       6.0      0.0     1665.0  ...   
2             0.0       8.0      0.0    13857.0  ...   
3             4.0      17.0      0.0     6216.0  ...   
4             0.0      13.0      0.0    12681.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    1                             0   
1                                    0                             0   
2                                    0                             0   
3                                    0                             1   
4                                    1                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                0   
1                            1                           0                1   
2                            1                           0                0   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                1                       1  
1                0                       1  
2                1                       1  
3                0                       1  
4                0                       1  

[5 rows x 91 columns]

In [59]:
# add missing dummy variables to testing set
for column in X_train.columns:
    if column not in X_test.columns:
        print(f'adding column {column}')
        X_test[column] = 0


adding column debt_settlement_flag_Y


    My prediction before creating, fitting, and scoring the models, and after looking at both of the datasets is that the Logistic Regression model will generate better results. 
    Even though both models used here (Logistic Regression, and Random Forest models) can be used on classification problems, with over 91 variables on this datasets, the Logistic model would perform better, as Random Forest is more for simpler classification problems.

In [60]:
# Train the Logistic Regression model on the unscaled data and print the model score
logisticRegr = LogisticRegression(
    solver='lbfgs', 
    max_iter=200,
    random_state=0).fit(X_train, y_train)
print('Logistic Regression train score: ', logisticRegr.score(X_train, y_train))
print('Logistic Regression test score: ', logisticRegr.score(X_test, y_test))

Logistic Regression train score:  0.6657635467980295
Logistic Regression test score:  0.525521054870268


/Users/elliotk/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [61]:
# Train a Random Forest Classifier model and print the model score
randomForestClass = RandomForestClassifier(random_state=0)
randomForestClass.fit(X_train, y_train)
print("RandomForestClassifier train score: ", randomForestClass.score(X_train,y_train))
print("RandomForestClassifier test score: ", randomForestClass.score(X_test,y_test))

RandomForestClassifier train score:  1.0
RandomForestClassifier test score:  0.6333475116971502


     After creating, fitting, and scoring the models, I was surprised that the Random Forest model performed better. This could be due to the fact that we are overfitting the model, or due to the fact that the data going into those models was never scaled, which is an important part in preprocessing. 

In [62]:
# Scale the data
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

    After scaling the data, and before re-fitting the models on the scaled data, my prediction remains the same. I still think that the Logistic Regression model will generate better results (for the same reasons as before).

In [63]:
# Train the Logistic Regression model on the scaled data and print the model score
logisticRegr_scaled = LogisticRegression(
    solver='lbfgs',
    max_iter=200,
    random_state=0
)
model = logisticRegr_scaled.fit(X_train_scaled, y_train)
print("LogisticRegression scaled train score: ", logisticRegr_scaled.score(X_train_scaled, y_train))
print("LogisticRegression scaled test score: ", logisticRegr_scaled.score(X_test_scaled, y_test))

LogisticRegression scaled train score:  0.7110016420361248
LogisticRegression scaled test score:  0.7603147596767333


/Users/elliotk/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [64]:
predictions = model.predict(X_test_scaled)
confusion_matrix(y_test, predictions)

array([[1764,  587],
       [ 540, 1811]])

In [65]:
# Train a Random Forest Classifier model on the scaled data and print the model score
randomForestClass_scaled = RandomForestClassifier(random_state=0)
model = randomForestClass_scaled.fit(X_train_scaled, y_train)
print("RandomForestClassifier scaled train score: ", randomForestClass_scaled.score(X_train_scaled, y_train))
print("RandomForestClassifier scaled test score: ", randomForestClass_scaled.score(X_test_scaled, y_test))

RandomForestClassifier scaled train score:  1.0
RandomForestClassifier scaled test score:  0.6344108889834114


In [66]:
predictions = model.predict(X_test_scaled)
confusion_matrix(y_test, predictions)

array([[1947,  404],
       [1315, 1036]])

    Once the data was scaled, and after re-fitting the models on the scaled data, the Logistic Regression was the better model. The Random Forest model remains overfitted, and even though its accuracy improved a tad (from 0.6333 to 0.6344), the Logistic Regression outperfomed it, going from 0.5255 to 0.7603.
    I still wouldn't feel comfortable using this model (accuracy of 0.76) to predict the credit risk of loans.  Specially after seeing the confusion matrix (on the Logistic Regression), ~11.48% of the testing data were false positives, and ~12.48% were false negatives, ~37.51% where true positives, and ~38.51% were true negatives. Money would be lost if we use this model for loan decisions.  